### ISSUE TITLE
Support "minimise to tray" on Linux

### ISSUE THREAD

**Issue by Betterbird** [Betterbird/thunderbird-patches#111]
From https://github.com/Betterbird/thunderbird-patches/issues/20#issuecomment-1383105403:

Check [this question](https://bugzilla.mozilla.org/show_bug.cgi?id=208923#c142) and [this answer](https://bugzilla.mozilla.org/show_bug.cgi?id=208923#c143). It's now tracked here:
https://bugzilla.mozilla.org/show_bug.cgi?id=1627479
Getting the indicator in the systray (Issue #20) was hard enough and still doesn't work fully on Gnome. Hiding the taskbar "button" to instead show the icon seems even harder. Or is there some API that can do it? Or maybe the code from this add-on https://addons.thunderbird.net/en-GB/thunderbird/addon/minimizetotray-reanimated/ could be used.

**Comment by Zahrun** [comment]
Yes, I actually got to know about Betterbird when I was looking for good alternatives for Thunderbird closing to tray on linux. MinimizeToTray Reanimated has version that works with Thunderbird 60.0 - 63.*, I think it is not maintained anymore. I used to use Birdtray, but had some issues with it. I am currently using SysTray-X, and it is fine on X11 but does not work on Wayland. Since now we have a tray icon that is working properly on Linux systems, I wish that Betterbird could support close to tray. It seems more attainable now. 

> I would love to have a "close to tray" option, and even nicer that a click on the tray icon would minimize to tray if betterbird is open and show betterbird if it was closed.

**Comment by Betterbird** [comment]
> How about adding a menu instead with a "Show/Hide" button? That was working out even for GNOME in the official appindicator demo.

Well, any additional text strings bring localisation issues. Two clicks instead of one are also clumsy and they are also different to what the Windows version does.

Yes, people have been adding menus to work around the non-ability of getting the clicks, see here for example:
https://bugzilla.redhat.com/show_bug.cgi?id=1319869
https://bugs.launchpad.net/ubuntu/+source/libappindicator/+bug/1910521
https://github.com/electron/electron/pull/35319/commits/f7adf8501fdd33b4b3c13649905dea977019eb59 <-- this shows the dummy menu.

We're in a better position: We use a fork of the library, so with the right knowledge (which we don't have) we could just add code for the click event.

**Comment by renyhp** [comment]
I landed here after installing Betterbird on linux and finding out there is still active development in trying to make a systray fully work (thanks!). I will just leave my comment here for anybody else that is in my same situation and immediately wants something that gives at least some basic functionality: [gyunaev/birdtray](https://github.com/gyunaev/birdtray) fully works with Betterbird, once you set the correct path to the betterbird executable and use `Betterbird` as the window pattern.

**Comment by Betterbird** [comment]
Yes, we have a patch that works in minimising the windows to the tray:
https://github.com/Betterbird/thunderbird-patches/blob/main/102/features/13-feature-linux-minimise-to-tray.patch
Sadly due to functionality missing from the upstream libayatana-appindicator, we can't get clicks onto the system tray icon to restore the window, so with the patch it's: Minimise once, not able to restore, kill the process 😭 

We made an effort here https://bugs.kde.org/show_bug.cgi?id=465438 to get some idea on how to implement that, so far to no avail. We need to start an initiative on one of the KDE mailing lists as suggested in the bug, [kde-devel@kde.org](mailto:kde-devel@kde.org).

**Comment by Zahrun** [comment]
I found that the Qt library provides the QSystemTrayIcon Class https://doc.qt.io/qt-6/qsystemtrayicon.html, which provides an icon for an application in the system tray.

> The QSystemTrayIcon class can be used on the following platforms:
      All supported versions of Windows.
      All Linux desktop environments that implement the D-Bus [StatusNotifierItem specification](http://www.freedesktop.org/wiki/Specifications/StatusNotifierItem/StatusNotifierItem), including KDE, Gnome, Xfce, LXQt, and DDE.
      All window managers and independent tray implementations for X11 that implement the [freedesktop.org XEmbed system tray specification](http://standards.freedesktop.org/systemtray-spec/systemtray-spec-0.2.html).
      All supported versions of macOS.

It has a "activated" signal which gives an "ActivationReason", which can be of type "QSystemTrayIcon::Trigger": The system tray entry was clicked

**Comment by rauldipeas** [comment]
@Betterbird do You know the [KDocker](https://github.com/user-none/KDocker)?

I'm using it to provide a tray icon to Betterbird and Thunderbird.

Maybe You can use something from the KDocker code to provide a decent solution.

**Comment by Betterbird** [comment]
Thanks for the suggestion. We're using AyatanaIndicators/libayatana-appindicator, and that's not moving forward, see latest comment here:
https://github.com/AyatanaIndicators/libayatana-appindicator/issues/4#issuecomment-1539138232

**Comment by Zahrun** [comment]
> @Betterbird do You know the [KDocker](https://github.com/user-none/KDocker)?
> 
> I'm using it to provide a tray icon to Betterbird and Thunderbird.
> 
> Maybe You can use something from the KDocker code to provide a decent solution.

KDocker does not support Wayland atm, so it’s not very future-proof. I would rather repeat my suggestion: QSystemTrayIcon. I understand that it would take some work to replace the existing code that uses libayatana-appindicator though.

**Comment by Betterbird** [comment]
Activation support for system tray icons has arrived, see https://github.com/AyatanaIndicators/libayatana-appindicator/pull/71.

Sadly it's only working for KDE, so we can implement "minimise to tray" on KDE only. Thoughts? Anyone willing to test?

**Comment by Zahrun** [comment]
I'm willing to test this on KDE

**Comment by Betterbird** [comment]
Thanks, I'll get it done in the next 48 hours. We already have a patch that is 95% complete:
https://github.com/Betterbird/thunderbird-patches/blob/main/102/features/13-feature-linux-minimise-to-tray.patch

**Comment by Betterbird** [comment]
OK, like on Windows, you can now enable this under "System Integration":

![image](https://github.com/Betterbird/thunderbird-patches/assets/86832723/d651bcbd-441f-496d-911d-bb2d5fe77d5d)

We tested the minimise part, we can't test the restore part since we currently don't have KDE. It's possible that it crashes on restore, that is, click onto the system tray icon after minimising it. If it's not minimised, a click onto the system tray icon should just bring the window to the foreground.

Please run Betterbird from a terminal window, if it doesn't even mininise to tray, it will tell you why.
Download: http://www.betterbird.eu/downloads2/betterbird-115.1.1-bb9-minimise-to-tray.en-US.linux-x86_64.tar.bz2

**Comment by Zahrun** [comment]
Thank you for that. It works. The only catch is that I have to click twice on the icon before the window is actually restored. It seems that the first time it tries to restore but then immediately goes back to tray. The second time, the window is restored fine.
https://github.com/Betterbird/thunderbird-patches/assets/10415894/59ae1590-bf2d-4f5f-967f-42cbd8f4966d

**Comment by Betterbird** [comment]
Thanks for testing. I can't see the "first time it tries to restore but then immediately goes back to tray", maybe it's not captured in the video.

There was a question you didn't answer: If it's not minimised, a click onto the system tray icon should just bring the window to the foreground. But I realise that for testing that, you need to set pref mail.biff.show_tray_icon_always first.

Looks like I have to get a KDE desktop somehow 😞.

**Comment by Betterbird** [comment]
Try:
http://www.betterbird.eu/downloads2/betterbird-115.1.1-bb9-minimise-to-tray-take2.en-US.linux-x86_64.tar.bz2

**Comment by Zahrun** [comment]
It works perfectly in KDE on X11 and Wayland. Just one question: is there an option to start betterbird minimized in the tray?

**Comment by Betterbird** [comment]
Yes, set pref mail.startupMinimized. In our testing that also works, but we've just made another tweak to make it more stable. Let me know if it doesn't work and I'll send another binary.

**Comment by felixoi** [comment]
I'll also leave my 2 cents here. Disclaimer: I know it should only work for KDE but I tried it with GNOME explicitly (by adjusting the config value). Even if there are no tooltips for GNOME I'd like to have this as feature.
This first take was very glitchy and not reliable at all. This one has reproducible behavior tho.

Here is a desktop recording:
https://github.com/Betterbird/thunderbird-patches/assets/14966472/edaf9fee-258d-4204-bf67-5e123fbe3f8a

1. Opening the application back up from the tray somehow opens it but in background (only visible in the taskbar). I'll have to actually select it in the taskbar again so it's shown. 
2. Only the first time it's working with one click, after having it opened back from tray once I'll have to double click the tray icon for the next times
3. Seems like it's always opening to the same location on the screen, not the one selected before

**Comment by Betterbird** [comment]
Gotta love those different desktops. Did the first version ever restore the window without having to click on the taskbar? We changed the code to restore the window, the first code was more Gtk-ish, so perhaps this works better.

The aim is to ship this in 115.2.0 in a few days, so there's not much time for experiments. I can compile a "take 3" later on for you to check.

**Comment by felixoi** [comment]
The first one was very unreliable as I said, so I can't really tell you what it did when 😄
Yes I just replaced "kde" with "gnome" in my case as I didn't know if adding works with commas or something different.

**Comment by Betterbird** [comment]
Commas or spaces to separate, or replace. But the console message said to add "gnome"?

**Comment by felixoi** [comment]
Yes, just re-checked by removing it again.
```
Betterbird: Window NOT minimized to system tray!
            You are using desktop environment gnome.
            Minimize to system tray is only supported for these desktops:
            kde.
            As an experiment, add gnome to preference
            mail.minimizeToTray.supportedDesktops.
            You might not be able to maximize the window again and
            will have to kill Betterbird.
            USE AT YOUR OWN RISK! If successful, contact support@betterbird.eu
```

**Comment by Betterbird** [comment]
FYI, there is code in Mozilla that checks for gnome and it's different to checking kde:
https://searchfox.org/mozilla-central/rev/7042a4a444f040ed87045c216f97cb73ce9410df/widget/gtk/WidgetUtilsGtk.cpp#485,491
For KDE they compare to "kde", for Gnome they search the string "gnome" in the desktop environment, as if the system returned "gnomeXYZ". Anyway, mostly irrelevant here.

Now I've added this code:
```
    void* gtkWidget = widget->GetNativeData(NS_NATIVE_SHELLWIDGET);
    if (gtkWidget) gtk_window_present(GTK_WINDOW(gtkWidget));
    widget->SetFocus(nsIWidget::Raise::Yes, mozilla::dom::CallerType::System);
```

The last take 1 had the first two lines, take 2 the last one, take 3 has all of them. Still works on KDE. I don't know what else to do if the windows manager doesn't react to the appropriate calls.

Take 3:
http://www.betterbird.eu/downloads2/betterbird-115.1.1-bb9-minimise-to-tray-take3.en-US.linux-x86_64.tar.bz2

If that works, I can add "gnome" to the pref and advertise the feature for KDE and Gnome.

**Comment by Betterbird** [comment]
Instead of `gtk_window_present()` https://docs.gtk.org/gtk4/method.Window.present.html, we can also try `gtk_window_deiconify()` https://docs.gtk.org/gtk3/method.Window.deiconify.html (or both):
https://searchfox.org/mozilla-central/rev/7042a4a444f040ed87045c216f97cb73ce9410df/widget/gtk/nsWindow.cpp#2832

Take 4:
http://www.betterbird.eu/downloads2/betterbird-115.1.1-bb9-minimise-to-tray-take4.en-US.linux-x86_64.tar.bz2

So this has "deiconify".

@felixoi, @Zahrun: Do you see any difference between take 2, 3 and 4? Help select the winner, unless they are all losers on Gnome.

**Comment by Betterbird** [comment]
I installed `sudo apt install vanilla-gnome-desktop`. So where do I get the "system tray" from? Following https://linuxiac.com/how-to-enable-system-tray-icons-in-gnome/, I installed the "Extension Manager" and then "AppIndicator and KStatusNotifierItem Support". Was that right? For activation of BB I had to triple click and all the rest was also very flaky. The Gnome I saw also didn't have a taskbar with icons like in your video.

**Comment by felixoi** [comment]
I'm using https://extensions.gnome.org/extension/1160/dash-to-panel/ which is pretty popular. For Manjaro for example it's the default extension if you want to have a toolbar.

Take 3 works the best for my GNOME installation currently. These points still stand tho:
1. Only the first time it's working with one click, after having it opened back from tray once I'll have to double click the tray icon for the next times
2. Seems like it's always opening to the same location on the screen, not the one selected before

The application opens back in the foreground with take3 also. With take4 its broken again.

**Comment by Zahrun** [comment]
> Yes, set pref mail.startupMinimized (no UI for that one). In our testing that also works, but we've just made another tweak to make it more stable. Let me know if it doesn't work and I'll send another binary.

Awesome.

> @felixoi, @Zahrun: Do you see any difference between take 2, 3 and 4? Help select the winner, unless they are all losers on Gnome.

I don’t see any difference between take 2, 3, and 4 on KDE.

> Take 3 works the best for my GNOME installation currently. These points still stand tho:
> 
>     1. Only the first time it's working with one click, after having it opened back from tray once I'll have to double click the tray icon for the next times
> 
>     2. Seems like it's always opening to the same location on the screen, not the one selected before

Point 2. also happens in KDE. The window gets restored to the same size but at the center of the screen instead of at the previous location. Personally I don’t mind.

I also noticed that when Betterbird is minimised to tray, if I call betterbird from Latte dock or simply by the application launcher (which is equivalent to just run the betterbird executable), it does not get restored. Maybe that would be a nice additional feature.

Thanks again for implementing this long awaited feature. Works pretty great on KDE.

**Comment by Betterbird** [comment]
OK, I went back to the code from "take 3" on Gnome, and it works pretty well. I can't detect any activity with a single click, for me to activate is always via double-click. So I'll ship with the code from "take 3" and gnome enabled by default.

**Comment by Betterbird** [comment]
Also see out "expert tips" page now:

<html><body>
<h3>System Tray support (Windows and Linux)</h3>
<p>The following is supported:</p>


OS/Desktop | System tray | System tray always | Tooltip with folder list | Minimize to tray
-- | -- | -- | -- | --
Windows | Yes | Yes | Yes | Yes
Linux/KDE | Yes | Yes | Yes | Yes
Linux/Gnome | Yes | Yes | No | Yes
Linux/Xfce | Yes | Yes | Yes | No


<small>"Minimize to tray" on Linux from version 115.2.0.</small>
<p>Relevant preferences:</p>
<ul>
<li>System tray: <tt>mail.biff.show_tray_icon</tt> (has UI)</li>
<li>System tray always: <tt>mail.biff.show_tray_icon_always</tt> (no UI)</li>
<li>Minimize to tray: <tt>mail.minimizeToTray</tt> (has UI), on Linux also: <tt>mail.minimizeToTray.supportedDesktops</tt> (no UI)</li>
<li>Start minimized: <tt>mail.startupMinimized</tt> (no UI)</li>
<li>Show new count: <tt>mail.biff.use_new_count_in_badge</tt> (has UI)</li>
</ul>
</body>
</html>

**Comment by dark-penguin** [comment]
I'm trying to understand why does it not work on MATE - this is the game-breaker that prevents me from upgrading from Thunderbird 45.

> UPDATE: Cinnamon and Mate also don't support the necessary activation for "minimize to tray".

You mean that the particular library you are using (`libayatana`?) does not catch events from Marco, correct? Clicking tray icons works just fine with any other app, including old Thunderbird with Firetray. I assume that's because they are using a different library, is that right? Could that maybe be used as a reference?

**Comment by Betterbird** [comment]
Please follow
https://github.com/AyatanaIndicators/libayatana-appindicator/pull/71

It's working now for KDE, Gnome and Xfce.

**Comment by dark-penguin** [comment]
Solution found! In https://github.com/mate-desktop/mate-panel/issues/1049 .
- Enable SNI support: `gsettings set org.mate.panel enable-sni-support true`
- Restart, because the panel can not pick up this setting at runtime
Now clicking the icon does restore the window!

If your Betterbird refuses to minimize to tray because it suspects that you won't be able to get it back (check its output console), then add `mate` to `mail.minimizeToTray.supportedDesktops` .

**Comment by Betterbird** [comment]
Thanks, so we've added this to the documentation:
<html><body>
<h3>System Tray support (Windows and Linux)</h3>
<p>The following is supported:</p>


OS / Desktop | System tray | System tray always | Tooltip with folder list | Minimize to tray
-- | -- | -- | -- | --
Windows | Yes | Yes | Yes | Yes
Linux / KDE | Yes | Yes | Yes | Yes
Linux / Gnome | Yes | Yes | No | Yes
Linux / Cinnamon | Yes | Yes | Yes | No
Linux / Mate | Yes | Yes | Yes | Yes(*)
Linux / Xfce | Yes | Yes | Yes | Yes


<small>"Minimize to tray" on Linux from version 102.5.0 and 115.2.0. 102.5.0 only supports KDE and Gnome.</small>
<p>(*) For the Mate desktop you need to run <tt>gsettings set org.mate.panel enable-sni-support true</tt> and then restart.
In Betterbird you need to add <tt>mate</tt> to the preference <tt>mail.minimizeToTray.supportedDesktops</tt>.
</p>
</body>
</html>

**Comment by Zahrun** [comment]
Thank you for your work, @Betterbird. I now completely replaced thunderbird by betterbird on my system, since the former lacked the minimise to tray feature for too many years. I sent a small donation to you via SEPA.

**Comment by githubuser6000** [comment]
Hey there, 
new betterbird user on Fedora KDE 39. I am using the flatpak. I cannot get minimize to tray working. 
After I hit minimize, I still see the app icon in the task manager. I did a reinstall, but now the the tray icon never shows, even when it is checked. 

How do I debug this?

**Comment by Betterbird** [comment]
> After I hit minimize, I still see the app icon in the task manager.

OK, so the "task manager" is what Windows calls "task bar":
https://userbase.kde.org/Plasma/Tasks
And this is set?
![image](https://github.com/Betterbird/thunderbird-patches/assets/86832723/aebeccdf-b33e-401d-a513-0575752f4ca6)

> I did a reinstall, but now the the tray icon never shows, even when it is checked.

What did you re-install? What is checked? The option to show the tray icon:
![image](https://github.com/Betterbird/thunderbird-patches/assets/86832723/a48a217a-d1bd-4c34-a277-7822760b6699)

What happens if you set pref `mail.biff.show_tray_icon_always`?

**Comment by githubuser6000** [comment]
> What did you re-install

I reinstalled betterbird from flatpak store

![image](https://github.com/Betterbird/thunderbird-patches/assets/111274694/627f110d-bfd4-4662-a46b-3c06ce94c132)
![image](https://github.com/Betterbird/thunderbird-patches/assets/111274694/fd639bdd-4800-4a2a-b9fa-b831d650b256)

When maximized left is the icons-only task manager thingy, and the right-most one is the tray icon. I edited the picture to remove the empty space in between.
![image](https://github.com/Betterbird/thunderbird-patches/assets/111274694/953c0561-09cb-4efc-839f-bb27773223b8)

This is betterbird minimized 
![image](https://github.com/Betterbird/thunderbird-patches/assets/111274694/0b3008e6-17e6-40e6-bde4-36ec8338dc7e)

> mail.biff.show_tray_icon_always
I set that to true, and I can see the tray icon when minimized or maximized. Clicking on the icon does not maximize the window when it is minimized. When minimized, both the tray icon and task manager icons are shown (the 4th image)

**Comment by githubuser6000** [comment]
> What did you re-install

I reinstalled betterbird from flatpak store

![image](https://github.com/Betterbird/thunderbird-patches/assets/111274694/627f110d-bfd4-4662-a46b-3c06ce94c132)
![image](https://github.com/Betterbird/thunderbird-patches/assets/111274694/fd639bdd-4800-4a2a-b9fa-b831d650b256)

When maximized left is the icons-only task manager thingy, and the right-most one is the tray icon. I edited the picture to remove the empty space in between.
![image](https://github.com/Betterbird/thunderbird-patches/assets/111274694/953c0561-09cb-4efc-839f-bb27773223b8)

This is betterbird minimized 
![image](https://github.com/Betterbird/thunderbird-patches/assets/111274694/0b3008e6-17e6-40e6-bde4-36ec8338dc7e)

> mail.biff.show_tray_icon_always

I set that to true, and I can see the tray icon when minimized or maximized. Clicking on the icon does not maximize the window when it is minimized. When minimized, both the tray icon and task manager icons are shown (the 4th image)

---

I downloaded the tarball so that all other variables are removed. I also deleted the `~/.thunderbird` folder. Then ran `betterbird`, and `betterbird-bin`. When X11 is used, minimizing the app correctly removes it from the task manager and shows the tray icon. 

With wayland enabled with `MOZ_ENABLE_WAYLAND=1`, I am able to reproduce the above issue. I want to use wayland because of better trackpad gesture support.

**Comment by Betterbird** [comment]
If I see it correctly, Wayland isn't fully supported by the Mozilla platform, see all the "Depends on" in this meta bug:
https://bugzilla.mozilla.org/show_bug.cgi?id=635134

That the window doesn't minimise is due to the fact that internally a "sizemodechange" is not received. We went debugged this with another BB user, who in the end reported this upstream bug:
https://bugzilla.mozilla.org/show_bug.cgi?id=1872790

You can repeat the steps in the description of this bug, that is, paste
```top.window.addEventListener("sizemodechange", () => console.log("sizemodechange received"))```
into the error console and see what gets logged when minimising the window.

**Comment by githubuser6000** [comment]
> top.window.addEventListener("sizemodechange", () => console.log("sizemodechange received"))

Yeah it doesn't print anything :/

**Comment by Betterbird** [comment]
Do you use Wayland? Anyway, TB doesn't offer the feature.

**Comment by lockie** [comment]
Nope, plain old X11.
I know, but neither does BB.

**Comment by Betterbird** [comment]
What you write is rather irritating. Yes, BB does offer system tray support incl. minimise to tray on known configurations and it's also documented at https://www.betterbird.eu/expert-tips/

<html><body>
<h3 id="system-tray">System Tray support (Windows and Linux)</h3>
<p>The following is supported:</p>

OS / Desktop | System tray | System tray always | Tooltip with folder list | Minimize to tray
-- | -- | -- | -- | --
Windows | Yes | Yes | Yes | Yes
Linux / KDE | Yes | Yes | Yes | Yes
Linux / Gnome | Yes | Yes | No | Yes(*)
Linux / Cinnamon | Yes | Yes | Yes | No
Linux / Mate | Yes | Yes | Yes | Yes(**)
Linux / Xfce | Yes | Yes | Yes | Yes


<small>"Minimize to tray" on Linux from version 102.5.0 and 115.2.0. 102.5.0 only supports KDE and Gnome.</small>
<p>(*) For Gnome you need to install the "AppIndicator and KStatusNotifierItem Support" extension. Double-click to activate.<br>
(**) The Mate desktop is supported from BB 115.3.1-bb14 build after 2 October 2023.</p>
<p>Relevant preferences:</p>
<ul>
<li>System tray: <tt>mail.biff.show_tray_icon</tt> (has UI)</li>
<li>System tray always: <tt>mail.biff.show_tray_icon_always</tt> (no UI)</li>
<li>Minimize to tray: <tt>mail.minimizeToTray</tt> (has UI)</li>
<li>Linux only: <tt>mail.minimizeToTray.supportedDesktops</tt> ("kde,gnome,xfce,mate") and 
                <tt>mail.minimizeToTray.desktopsGtkWindowPresentWithTime</tt> (special activation, "mate") (no UI)</li>
<li>Start minimized: <tt>mail.startupMinimized</tt> (no UI)</li>
<li>Windows only: Show count in taskbar badge: <tt>mail.biff.show_badge</tt> (has UI)</li>
<li>Show new count: <tt>mail.biff.use_new_count_in_badge</tt> (has UI)</li>
<li>Only count unread message in inboxes: <tt>mail.notification.count.inbox_only</tt> (no UI)</li>
</ul>
</body>
</html>

You will understand that it's impossible to support all possible Linux variations.

Which desktop are you using and have you started BB from the command line to see which desktop it reports. If that's not in the list of supported desktops, it won't work.

**Comment by dark-penguin** [comment]
@Betterbird He just means, "Since it's not working for me, and this is the only feature I needed, I'll stick with the unpatched Thunderbird for now". :slightly_smiling_face:

**Comment by Betterbird** [comment]
I know what he means. But it's written as a threat "back to Thunderbird then" and a false statement "neither does BB".

**Comment by lockie** [comment]
@Betterbird  I'm sorry I've offended you, I was just stating the fact that I'm not going to use your product since it does not work for me and, I guess, for a whole lot of professional Linux users since tiling window managers occupy a lot of places in the lists of most popular desktop environments, e.g. [this one](https://www.slant.co/topics/14976/~desktop-environments-for-arch-linux) — it even has one on 2nd place.
Is it ok for me to open separate issue regarding tray icon on tiling window managers and help with resolving it?

**Comment by Betterbird** [comment]
First of all, there is some confusion here. For system tray and minimise to tray to work, your Linux desktop needs to provide a system tray. As you can see in https://github.com/Betterbird/thunderbird-patches/issues/111#issuecomment-2138996896, some desktops support this, Gnome only after an additional software installation.

Reading https://askubuntu.com/questions/18078/what-is-the-difference-between-a-desktop-environment-and-a-window-manager, there is a difference between desktop and windows manager. We're running Xfce with xfwm4, KDE with KWin, Mate with Metacity and Gnome with "GNOME Shell".

The system tray support depends on the desktop environment. That's why BB on Linux, when run from the command, line reports the DE is has detected.

Why don't you read the comment above again, run BB from the command line and see which DE it reports? You can add the DE name string to pref `mail.minimizeToTray.supportedDesktops` and see whether it works.

BTW, your comment https://github.com/Betterbird/thunderbird-patches/issues/111#issuecomment-2138845801 is unclear. Which icon do you click? The close button will always close the application. And minimise to tray needs to be enabled via an option.

Over to you to provide the necessary details instead of a "doesn't work".

**Comment by dark-penguin** [comment]
@lockie Tray functionality sucks because it's implemented via GTK3 - and we can't change this because that's what Thunderbird uses. And GTK3 is that famous one when the Gnome devs decided that they have their vision for the future of desktop computing, and that future does not have trays at all (and made Gnome 3). :face_with_spiral_eyes: So the only way to have an icon is to use something called libayatana-appindicator, which does not make things easy either. And on top of all that, each DE has their own tray implementation... and there are multiple different protocols...

If you are curious to see what we had to go through to make this work at all, and pick up where we left, there are a few issues around the time of this one (numbered something about 100-111 IIRC). There's even a minimal test source to fiddle around with. There are a couple different issues for different DEs, so we had to implement an option to switch between the behaviour that doesn't work on KDE and the one that doesn't work on MATE. :dizzy_face: So if you can figure out how to get tray icons working in your DE, and it doesn't break the other ones too much, then it could potentially be added as well. (I'm not a core dev though - I just did the same for MATE to get it working for me. :slightly_smiling_face: )

**Comment by Betterbird** [comment]
Right.

Since competition is good for business, the TB folks have now started to implement something themselves:
https://bugzilla.mozilla.org/show_bug.cgi?id=18732
https://phabricator.services.mozilla.com/D211940
https://github.com/thunderbird/linux-sys-tray

Good luck to them.

**Comment by lockie** [comment]
Right, I wasn't providing enough details. I'm using polybar panel to [display a system tray](https://polybar.readthedocs.io/en/latest/user/modules/tray.html). Betterbird correctly identifies my desktop as xmonad, and I already have added it to `mail.minimizeToTray.supportedDesktops`. Thing is, tiling window managers do not have the concept of "minimizing" windows per se, so it'd be nice if BB would minimize by clicking on the icon. Or perhaps minimize to tray instead of closing when the window gets the close event.

**Comment by dark-penguin** [comment]
Then that's the problem. @Betterbird refuses to

**Comment by dark-penguin** [comment]
> Thing is, tiling window managers do not have the concept of "minimizing" windows per se

Then that's the problem. @Betterbird refuses to make it minimize on tray icon click, in order to keep the behaviour consistent with Windows, so if you don't have a way to issue a "minimize" command normally, there's no other way to do it.

I do remember some discussion about minimizing instead of closing, but I don't remember @Betterbird 's position on it...

**Comment by Betterbird** [comment]
Hmm, not enough details. You haven't stated whether "minimise to tray" is actually selected as an option:
![image](https://github.com/Betterbird/thunderbird-patches/assets/86832723/34d5044c-0767-4955-aa12-c0fce62fbcfc)

You've added `xmonad` to pref `mail.minimizeToTray.supportedDesktops` and then what happened? But you're saying that your wm doesn't have a concept of minimising? How do you want to minimise if the wm doesn't support it? Behind the scenes, we're using libayatana-appindicator and that's doing a bunch of Gtk and dbus calls at the backend. If the receiving wm or DE doesn't do the "right thing", we won't be able to fix it.

Next you talk about "the icon", but there are several, at least one to minimise and one to closes.

With the add-on https://addons.thunderbird.net/en-us/thunderbird/addon/minimize-on-close/ you can change the action of the close icon, the app will minimise instead, but then again, I don't know whether minimise means anything in you wm.

**Comment by doronbehar** [comment]
I see that this thread is old, but a alive. I too can't get system tray support to work. Here are my settings:

![image](https://github.com/Betterbird/thunderbird-patches/assets/10998835/55c8d9e7-1c37-46e2-9adc-72eeeef9bc23)

Any my betterbird version:

![image](https://github.com/Betterbird/thunderbird-patches/assets/10998835/cb90c879-1ad8-47a8-bf9e-1a76c7526608)

I'm using NixOS (Thanks to `@SuperSandro2000` who packaged it for us). I also have a full build log of Betterbird to share (sharing it so you could maybe find a patch we are missing or something alike):

https://cache.nixos.org/log/q6cw2g05v27jypqyni60rpnzvdjsp11i-betterbird-unwrapped-115.9.0.drv

On the contrary, the flatpak version of this does show the systray, and together with [minimize-on-close](https://addons.thunderbird.net/en-US/thunderbird/addon/minimize-on-close/) the experience is perfect.

Still I'd like to get this working natively on NixOS.

**Comment by Betterbird** [comment]
I think that if the FlatPak works but the NixOS build doesn't then that wasn't built correctly. Hard to say. Have you tried our binary release (.tar.bz2 files)?

Looks like these patches
12-feature-linux-systray.patch
12-feature-linux-systray-compile.patch
12-feature-linux-systray-tooltip.patch
12a-feature-linux-systray-tooltip.patch
12-feature-linux-systray-no-root.patch
12-feature-linux-systray-example.patch
13-feature-linux-systray-activate.patch
13-feature-linux-systray-activate-xfce.patch
13-feature-linux-systray-activate-mate.patch
12+13-feature-linux-systray-incl-minimise.patch
were successfully applied. It's strange though that these lines occur:

```
Checking patch third_party/appindicator/app-indicator.c...
Hunk #1 succeeded at 125 (offset 1 line).
Hunk #2 succeeded at 708 (offset 1 line).
Hunk #3 succeeded at 1366 (offset 2 lines)
```

I don't know how the patches are applied, let's assume the patches all applied and only the process is more noisy.

**Comment by doronbehar** [comment]
@Betterbird I suspect that this is related to Ayatana dependency not available during the build. For sure our package `libayatana-appindicator` (built from [this source](https://github.com/AyatanaIndicators/libayatana-appindicator
)), is not added to the build. If this is indeed the reason. I'm surprised that the build didn't fail due to this missing dependency. I'm also wondering why didn't the build log mention even a search for this dependency.

**Comment by Betterbird** [comment]
No. We include the library ourselves.

**Comment by doronbehar** [comment]
> I think that if the FlatPak works but the NixOS build doesn't then that wasn't built correctly. Hard to say.
> ...
> It's strange though that these lines occur:
> 
> ```
> Checking patch third_party/appindicator/app-indicator.c...
> Hunk #1 succeeded at 125 (offset 1 line).
> Hunk #2 succeeded at 708 (offset 1 line).
> Hunk #3 succeeded at 1366 (offset 2 lines)
> ```

This is how we do it:

https://github.com/NixOS/nixpkgs/blob/9b19ff75c0338513ee5732ab8ebfd58ac7e5a770/pkgs/applications/networking/mailreaders/betterbird/default.nix#L56-L95

**Comment by Betterbird** [comment]
OK, the `git apply -v` makes it more noisy. I'm not thrilled that you remove 14-feature-regexp-searchterm.patch. In the current version there are more patches relying on ICU.

Other than that, I can't tell why the system tray stuff doesn't work.

**Comment by doronbehar** [comment]
> OK, the `git apply -v` makes it more noisy. I'm not thrilled that you remove 14-feature-regexp-searchterm.patch. In the current version there are more patches relying on ICU.
> 
> Other than that, I can't tell why the system tray stuff doesn't work.

Maybe @SuperSandro2000 will have a clue :pray:.

**Comment by MarsAstro** [comment]
The minimize option doesn't seem to work right on Linux Mint 22 Cinnamon edition either. I've installed Betterbird via the Software Manager (Flathub), and with the option on minimizing leaves the app on the bar instead of moving it to tray.

**Comment by Betterbird** [comment]
Cinnamon doesn't work:
http://www.betterbird.eu/expert-tips/

Thanks for confirming.

**Comment by MarsAstro** [comment]
> Cinnamon doesn't work: http://www.betterbird.eu/expert-tips/ Thanks for confirming.

Ah, I see! Is this a limitation with the distro, or is it something that could potentially be fixed in future updates?

**Comment by Betterbird** [comment]
No intention to work around quirks of the various desktops. It works for 
KDE, Gnome, Xfce and Mate, and Mate was already a pain to get working.

Besides, the TB folks have just implemented their own system tray support 
written in Rust:
https://github.com/thunderbird/linux-sys-tray
https://github.com/thunderbird/dbus_hooks

Let's see how well this works, it only mentions KDE and Gnome.

Ideally we can scrap the Betterbird implementation at some stage.

**Comment by code-IM-perfect** [comment]
For any awesomewm users (although I believe it should work on any Window Manager that supports System Tray), it works but you need to add 'awesome' to `mail.minimizeToTray.supportedDesktops` in the Config Editor

Config editor can be opened with Settings > General > Config Editor

**Comment by serious-angel** [comment]
Holy gracious smokes! Apparently, there's some progress, and though no Linux yet [implemented](<https://bugzilla.mozilla.org/show_bug.cgi?id=1942125>), at least there's hope!  

> If you use Thunderbird on Windows, you may be familiar with sending Thunderbird to the system tray (the notification area) so it keeps running in the background. Starting with Thunderbird 154, the way this works has changed – instead of minimizing to the tray, Thunderbird now moves to the tray when you close it. This article explains the change and shows you how to turn the feature on.
> 
> Note: This feature applies only to Thunderbird 154 and later on Windows 10 and 11. It is not available on macOS or Linux.
> 
> Source: <https://support.mozilla.org/en-US/kb/close-tray-starting-154>

Though, no support for Linux, the OS I've been into for more than two decades now, it at least has a progress...  

> Honestly the only thing holding up progress is an engineers time to work on it. I started this project some time ago and then would love to find time to come back to it.  
> [Source](<https://bugzilla.mozilla.org/show_bug.cgi?id=1942125#c12>)

Thank you... heartfelt...

**Comment by Betterbird** [comment]
You're living under a rock, right?

https://blog.betterbird.eu/2026/06/close-to-tray-is-coming-the-betterbird-153

**Comment by serious-angel** [comment]
> You're living under a rock, right?
> 
> https://blog.betterbird.eu/2026/06/close-to-tray-is-coming-the-betterbird-153

Thank you for sending a person under the rock, for an article posted on `2026-06-24`, or fairly few days ago.  

I've added you to the blocked on my account, and it's sorrowful to know such toxic people as you.  
You are free to block me, too, if you hate people so much, including those who try to participate whenever they can with their own local duties and lives.  

I deleted your project, I used for a ~week, from personal environments, and won't recommend your project to anyone I know since this moment, and I regret investing my time into it, and this issue at this point.  

It's quite shocking to know you act so, sorry, awfully and disrespectfully framing and devaluing such a project you represent, too... 

Such a... sorrow... and I had no idea, nor expected such a... response...  
And I am sorry if I did anything wrong, too, if anything...  
Stay safe...

**Comment by Betterbird** [comment]
Frankly, I don't understand why you polluted this thread with a comment on "hope for Thunderbird" when the feature is implemented in BB for years.

Then you're quoting a TB KB article, which post-dates our blog post. The KB article also doesn't apply to BB.

So, basically, you're uninformed and posting irrelevant content. And that supposedly merits a lovey-dovey reply. It does not.
